sdks/java/io/amqp/src/main/java/org/apache/beam/sdk/io/amqp/AmqpIO.java

```diff
@@ -0,0 +1,471 @@
+import org.apache.beam.sdk.transforms.display.DisplayData;
+import org.apache.beam.sdk.values.PBegin;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PDone;
+import org.apache.qpid.proton.amqp.messaging.AmqpValue;
+import org.apache.qpid.proton.message.Message;
+import org.apache.qpid.proton.messenger.Messenger;
+import org.apache.qpid.proton.messenger.Tracker;
+import org.joda.time.Duration;
+import org.joda.time.Instant;
+
+/**
+ * AmqpIO supports AMQP 1.0 protocol using the Apache QPid Proton-J library.
+ *
+ * <p>If you want to use AMQP 0.9 protocol, you might also be interested in the Apache Beam
+ * RabbitMqIO.
+ *
+ * <p>It's also possible to use AMQP 1.0 protocol via Apache Qpid JMS connection factory and the
+ * Apache Beam JmsIO.
+ *
+ * <h3>Binding AMQP and receive messages</h3>
+ *
+ * <p>The {@link AmqpIO} {@link Read} can bind a AMQP listener endpoint and receive messages. It can
+ * also connect to a AMPQ broker (such as Apache Qpid or Apache ActiveMQ).
+ *
+ * <p>{@link AmqpIO} {@link Read} returns an unbounded {@link PCollection} of {@code String}
+ * containing the received messages (as a {@link AmqpValue}).
+ *
+ * <p>To configure a AMQP source, you have to provide a list of addresses where it will receive
+ * messages. An address has the following form: {@code
+ * [amqp[s]://][user[:password]@]domain[/[name]]} where {@code domain} can be one of {@code
+ * host | host:port | ip | ip:port | name}. NB: the {@code ~} character allows to bind a AMQP
+ * listener instead of connecting to a remote broker. For instance {@code amqp://~0.0.0.0:1234}
+ * will bind a AMQP listener on any network interface on the 1234 port number.
+ *
+ * <p>The following example illustrates how to configure a AMQP source:
+ *
+ * <pre>{@code
+ *
+ *  pipeline.apply(AmqpIO.read()
+ *    .withAddresses(Collections.singletonList("amqp://host:1234")))
+ *
+ * }</pre>
+ *
+ * <h3>Sending messages to a AMQP endpoint</h3>
+ *
+ * <p>{@link AmqpIO} provides a sink to send {@link PCollection} elements as messages.
+ *
+ * <p>As for the {@link Read}, {@link AmqpIO} {@link Write} requires a list of addresses where to
+ * send messages. The following example illustrates how to configure the {@link AmqpIO}
+ * {@link Write}:
+ *
+ * <pre>{@code
+ *
+ *  pipeline
+ *    .apply(...) // provide PCollection<String>
+ *    .apply(AmqpIO.write()
+ *      .withAddresses(Collections.singletonList("amqp://host:1234")));
+ *
+ * }</pre>
+ */
+public class AmqpIO {
+
+  public static Read read() {
+    return new AutoValue_AmqpIO_Read.Builder().setMaxNumRecords(Long.MAX_VALUE).build();
+  }
+
+  public static Write write() {
+    return new AutoValue_AmqpIO_Write.Builder().build();
+  }
+
+  private AmqpIO() {
+  }
+
+  /**
+   * A {@link PTransform} to read/receive messages using AMQP 1.0 protocol.
+   */
+  @AutoValue
+  public abstract static class Read extends PTransform<PBegin, PCollection<String>> {
```

Comment:
Can this return a `PCollection<Message>`, just like Pubsub and Kafka do? Likewise for write.